### DISCUSSION TITLE
Retrieve count of an ORM select query

### DISCUSSION THREAD

**Discussion by MaartenUreel** [sqlalchemy/sqlalchemy#9659] [Usage Questions] [answered]
Assuming I have these two queries:

```python
messages = select(VoicemailMessage).where(VoicemailMessage.voicemail_box_id == self.box.id)
unplayed_messages = messages.where(VoicemailMessage.is_played == False)
played_messages = messages.where(VoicemailMessage.is_played == True)
```

Is it possible to retrieve the record count of these two queries before actually retrieving all the data?

Something alongs the lines of:
```python
# Perform COUNT here
if unplayed_messages.count() > 0:
   speak("These are your messages")

   # And actually retrieve here
   for m in self.x.db.scalars(unplayed_messages):
       self.play_message(m)
```

`self.x.db` is a Session instance here.

**Comment by zzzeek** [comment, marked as answer]
yes [replace the columns](https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Select.with_only_columns) with func.count() :

    count = db.scalar(unplayed_messages.with_only_columns(func.count(), maintain_column_froms=True))

**Reply by CaselIT** [reply]
Just note that the above may not be 100% correct on complex selects that include distinct / group_by. In these case using a subquery is probably better
```
count = db.scalar(select(funct.count(1).select_from(unplayed_messages.subquery())
```

**Reply by MaartenUreel** [reply]
Perfect, thank you for the clear explanations.
